go/perf/suite/suite.go

```diff
@@ -0,0 +1,334 @@
+// (perf) RUN(1/3) TestFoo (recorded as "Foo")
+// (perf) PASS:    TestFoo (10s, paused 1s, total 11s)
+// (perf) RUN(1/3) TestZoo (recorded as "Zoo")
+// (perf) PASS:    TestZoo (1s, paused 42s, total 43s)
+// ...
+//
+// > noms show http://localhost:8000::csv-import
+// {
+//   environment: ...
+//   tests: [{
+//     "Bar": {elapsed: 15s, paused: 2s,  total: 17s},
+//     "Foo": {elapsed: 10s, paused: 1s,  total: 11s},
+//     "Qux": {elapsed: 5s,  paused: 15s, total: 20s},
+//     "Zoo": {elapsed: 1s,  paused: 42s, total: 43s},
+//   }, ...]
+//   ...
+// }
+// ```
+package suite
+
+import (
+	"bytes"
+	"flag"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path"
+	"reflect"
+	"regexp"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/attic-labs/noms/go/chunks"
+	"github.com/attic-labs/noms/go/datas"
+	"github.com/attic-labs/noms/go/dataset"
+	"github.com/attic-labs/noms/go/marshal"
+	"github.com/attic-labs/noms/go/spec"
+	"github.com/attic-labs/noms/go/types"
+	"github.com/attic-labs/testify/assert"
+	"github.com/shirou/gopsutil/cpu"
+	"github.com/shirou/gopsutil/disk"
+	"github.com/shirou/gopsutil/host"
+	"github.com/shirou/gopsutil/mem"
+)
+
+var (
+	perfFlag        = flag.String("perf", "", "The database to write perf tests to. If this isn't specified, perf tests are skipped. If you want a dry run, use \"mem\" as a database")
+	perfRepeatFlag  = flag.Int("perf.repeat", 1, "The number of times to repeat each perf test")
+	testNamePattern = regexp.MustCompile("^Test[0-9]*([A-Z].*$)")
+)
+
+// PerfSuite is the core of the perf testing suite. See package documentation for details.
+type PerfSuite struct {
+	// T is the `testing.T` instance set when the suite is passed into `Run`.
+	T *testing.T
+
+	// W is the `io.Writer` to write test output, which only outputs if the verbose flag is set.
+	W io.Writer
+
+	// AtticLabs is the path to the attic-labs directory (e.g. /path/to/go/src/github.com/attic-labs).
+	AtticLabs string
+
+	// Database is a Noms database that tests can use for reading and writing. State is persisted across a single run of a suite.
+	Database datas.Database
+
+	// DatabaseSpec is the Noms spec of `Database` (typically a localhost URL).
+	DatabaseSpec string
+
+	tempFiles []*os.File
+	paused    time.Duration
+}
+
+type perfSuiteT interface {
+	Suite() *PerfSuite
+}
+
+type timeInfo struct {
+	elapsed, paused, total time.Duration
+}
+
+type testRun map[string]timeInfo
+
+type nopWriter struct{}
+
+func (r nopWriter) Write(p []byte) (int, error) {
+	return len(p), nil
+}
+
+// Run runs `suiteT` and writes results to dataset `datasetID` in the database given by the `-perf` command line flag.
+func Run(datasetID string, t *testing.T, suiteT perfSuiteT) {
+	if *perfFlag == "" {
+		return
+	}
+
+	assert := assert.New(t)
+
+	// Piggy-back off the go test -v flag.
+	verboseFlag := flag.Lookup("test.v")
+	assert.NotNil(verboseFlag)
+	verbose := verboseFlag.Value.(flag.Getter).Get().(bool)
+
+	suite := suiteT.Suite()
```

Comment:
People's perf suite does embed a PerfSuite (and inherits the Suite method). This interface is just a subclass hack so that Run works i.e. it can't take a suite.PerfSuite because that's a struct.

Testify did something similar-ish here: https://github.com/stretchr/testify/blob/master/suite/interfaces.go